fix: Can not sync job status correctly when upgrading from v1.5

QingyaFan opened the pull request:
v1.5 changed the naming logics of pod group by adding UID into the name: #2140, and there is also another fix regarding handling the already created pod group without UID in create or update: #2400. But a similar fix does not exist in the syncJob function.

Fixes #3640

Monokaix commented:
Welcome! Please sign off your commit with `git commit -s`.

Monokaix commented on pkg/controllers/job/job_controller_actions.go:
```diff
@@ -152,11 +152,18 @@ func (cc *jobcontroller) killJob(jobInfo *apis.JobInfo, podRetainPhase state.Pha
-	if err := cc.vcClient.SchedulingV1beta1().PodGroups(job.Namespace).Delete(context.TODO(), pgName, metav1.DeleteOptions{}); err != nil {
+	pgName := jobhelpers.GetRelatedPodGroupName(job)
+	pgIface := cc.vcClient.SchedulingV1beta1().PodGroups(job.Namespace)
+	if err := pgIface.Delete(context.TODO(), pgName, metav1.DeleteOptions{}); err != nil {
```
Why named Iface here?

Monokaix replied:
I mean pgIface itself is bit ambiguous: )

QingyaFan replied:
I noticed `cc.vcClient.SchedulingV1beta1().PodGroups(job.Namespace)` returns a `v1beta1.PodGroupInterface` type, so i set the variable name to pgIface(v1beta1.**P**od**G**roup**I**nter**face**), can you give me a suggestion 🙇🏼

Monokaix replied:
I think pgClient is ok.

QingyaFan commented:
/assign @hwdef

Monokaix commented:
There are many duplicate logic when get deal with legacy pg and codes are not so readable，I think we can wrap a func to get real pg and do not expose so many logic in where controller call it.

QingyaFan commented:
> There are many duplicate logic when dealing with legacy pg and codes are not so readable，I think we can wrap a func to get real pg and do not expose so many logic in where controller call it.

Yeah, that make sense. What you mean is wrap a func that check which pg(normal or legacy) exist in k8s cluster, then return it. However, this will cause a problem that: in race condition, when we get the pg name, and before we operate it, perhaps other routine has already operate it, for example maybe the pg has already been deleted. The get and the other operation are not in a transaction.

Monokaix commented:
> > There are many duplicate logic when dealing with legacy pg and codes are not so readable，I think we can wrap a func to get real pg and do not expose so many logic in where controller call it.
> 
> Yeah, that make sense. What you mean is wrap a func that check which pg(normal or legacy) exist in k8s cluster, then return it. However, this will cause a problem that: in race condition, when we get the pg name, and before we operate it, perhaps other routine has already operate it, for example maybe the pg has already been deleted. The get and the other operation are not in a transaction.

That's a good point, but seems that current code is also executed in order and no lock is added，right?

Monokaix commented:
> > There are many duplicate logic when dealing with legacy pg and codes are not so readable，I think we can wrap a func to get real pg and do not expose so many logic in where controller call it.
> 
> Yeah, that make sense. What you mean is wrap a func that check which pg(normal or legacy) exist in k8s cluster, then return it. However, this will cause a problem that: in race condition, when we get the pg name, and before we operate it, perhaps other routine has already operate it, for example maybe the pg has already been deleted. The get and the other operation are not in a transaction.

And if we get a pg and then call Delete method to delete the pg, if the pg is alredy deleted then NotFoundErr will also returned and we can be aware of it.

QingyaFan commented:
> > > There are many duplicate logic when dealing with legacy pg and codes are not so readable，I think we can wrap a func to get real pg and do not expose so many logic in where controller call it.
> > 
> > 
> > Yeah, that make sense. What you mean is wrap a func that check which pg(normal or legacy) exist in k8s cluster, then return it. However, this will cause a problem that: in race condition, when we get the pg name, and before we operate it, perhaps other routine has already operate it, for example maybe the pg has already been deleted. The get and the other operation are not in a transaction.
> 
> And if we get a pg and then call Delete method to delete the pg, if the pg is alredy deleted then NotFoundErr will also returned and we can be aware of it.

ok, I updated the code, please review again.

Monokaix commented on pkg/controllers/job/job_controller_actions.go:
```diff
@@ -42,6 +42,30 @@ import (
+// getRelatedPodGroup returns the podgroup related to the vcjob.
+// it will return normal pg if it exist in cluster,
+// else it return legacy pg before version 1.5.
+func (cc *jobcontroller) getRelatedPodGroup(job *batch.Job) (*scheduling.PodGroup, error) {
```
getRelatedPodGroup -》 getPodGroupByJob

Monokaix commented on pkg/controllers/job/job_controller_actions.go:
```diff
@@ -42,6 +42,30 @@ import (
+// it will return normal pg if it exist in cluster,
+// else it return legacy pg before version 1.5.
+func (cc *jobcontroller) getRelatedPodGroup(job *batch.Job) (*scheduling.PodGroup, error) {
+	pgName := cc.generateRelatedPodGroupName(job)
```
```go
	pgName := cc.generateRelatedPodGroupName(job)
	pg, err := cc.pgLister.PodGroups(job.Namespace).Get(pgName)
	if err == nil {
		return pg, nil
	}
	if apierrors.IsNotFound(err) {
		pg, err = cc.pgLister.PodGroups(job.Namespace).Get(job.Name)
		if err != nil {
			return nil, err
		}
		return pg, nil
	}
	return nil, err
Change to this is better: )

Monokaix commented on pkg/controllers/job/job_controller_actions.go:
```diff
@@ -152,12 +176,13 @@ func (cc *jobcontroller) killJob(jobInfo *apis.JobInfo, podRetainPhase state.Pha
-			klog.Errorf("Failed to delete PodGroup of Job %v/%v: %v",
-				job.Namespace, job.Name, err)
-			return err
+	pg, _ := cc.getRelatedPodGroup(job)
```
We should check the returned err first.

QingyaFan commented:
I improved the code according to your comment. Please take a look again :) @Monokaix

Monokaix commented:
/reopen

QingyaFan commented:
/reopen

QingyaFan commented:
I accidently close the PR when i rebase code just now, now I have solve the commit message problem, it's ok to continue. @Monokaix

Monokaix commented:
/lgtm

QingyaFan commented:
/retest

QingyaFan commented:
@Monokaix seems "E2E Sequence Jobs" failed, i have checked the output, can't find what's wrong, can you help me?

Monokaix commented:
> @Monokaix seems "E2E Sequence Jobs" failed, i have checked the output, can't find what's wrong, can you help me?

I will check that later.

Monokaix commented:
You can use `git commit --ame --no-edit -s` and for push again to retest.

QingyaFan commented:
> New changes are detected. LGTM label has been removed.

Please help add the lgtm label again to retest @Monokaix

Monokaix commented:
/ok-to-test

william-wang commented:
@QingyaFan please add ut for this change.

QingyaFan commented:
> @QingyaFan please add ut for this change.

ok，I will try to add ut.

QingyaFan commented:
> @QingyaFan please add ut for this change.

ut has been added, please check it out when you have time.

QingyaFan commented:
@Monokaix can you help me review the unit test code and merge my pull request, @william-wang seems have no time

QingyaFan commented:
Please help review the unit test code and merge the pull request @Monokaix @hwdef @hzxuzhonghu @william-wang

Monokaix commented:
line 205 in your pr of file `job_controller_actions_test.go` didn't check the expected err and actual err, we should modify it: )

QingyaFan commented:
> line 205 in your pr of file `job_controller_actions_test.go` didn't check the expected err and actual err, we should modify it: )

however，i did not change line 205...

Monokaix commented:
/ok-to-test

Monokaix commented:
HI，please rebase your pr: )

QingyaFan commented:
@Monokaix I fix the conflict. Can you add lgtm label. And the `Vcctl Test / E2E about Volcano CLI (pull_request)` test failed, can you retrigger the test to see if this happens occasionally(because the pr did not change the failed test file). Thanks !

hwdef commented:
> ok, I updated the code, please review again.

Please squash your commit to one, the CI will restart automatically

QingyaFan commented:
> > ok, I updated the code, please review again.
> 
> Please squash your commit to one, the CI will restart automatically

the commits are split in different branches and not continus, i don't know how to squash into one. what about i redo the change in a different branch and create a new pull request?

Monokaix commented:
> > > ok, I updated the code, please review again.
> > 
> > 
> > Please squash your commit to one, the CI will restart automatically
> 
> the commits are split in different branches and not continus, i don't know how to squash into one. what about i redo the change in a different branch and create a new pull request?

If you are not familiar with git，you can build a new branch such as `tmp` locally based on the current branch first, and reset your current branch to latest master，and then just cherry pick your own commit from the `tmp` branch.

Monokaix commented:
> > > > ok, I updated the code, please review again.
> > > 
> > > 
> > > Please squash your commit to one, the CI will restart automatically
> > 
> > 
> > the commits are split in different branches and not continus, i don't know how to squash into one. what about i redo the change in a different branch and create a new pull request?
> 
> If you are not familiar with git，you can build a new branch such as `tmp` locally based on the current branch first, and reset your current branch to latest master，and then just cherry pick your own commit from the `tmp` branch.


> > > ok, I updated the code, please review again.
> > 
> > 
> > Please squash your commit to one, the CI will restart automatically
> 
> the commits are split in different branches and not continus, i don't know how to squash into one. what about i redo the change in a different branch and create a new pull request?

You can do that: )

QingyaFan commented:
> > > > > ok, I updated the code, please review again.
> > > > 
> > > > 
> > > > Please squash your commit to one, the CI will restart automatically
> > > 
> > > 
> > > the commits are split in different branches and not continus, i don't know how to squash into one. what about i redo the change in a different branch and create a new pull request?
> > 
> > 
> > If you are not familiar with git，you can build a new branch such as `tmp` locally based on the current branch first, and reset your current branch to latest master，and then just cherry pick your own commit from the `tmp` branch.
> 
> > > > ok, I updated the code, please review again.
> > > 
> > > 
> > > Please squash your commit to one, the CI will restart automatically
> > 
> > 
> > the commits are split in different branches and not continus, i don't know how to squash into one. what about i redo the change in a different branch and create a new pull request?
> 
> You can do that: )

I create a new pull request, please have a look: https://github.com/volcano-sh/volcano/pull/3786 @Monokaix

Monokaix commented:
tracked by https://github.com/volcano-sh/volcano/pull/3786

Monokaix commented:
/close
